ext/drbbdup/drbbdup.c

```diff
@@ -383,8 +381,15 @@ drbbdup_duplicate_phase(void *drcontext, void *tag, instrlist_t *bb, bool for_tr
     drbbdup_manager_t *manager =
         (drbbdup_manager_t *)hashtable_lookup(&manager_table, pc);
 
-    /* A manager is created if there does not exit one that does not already
-     * book-keeping this bb. */
+    if (!for_trace && manager != NULL && !manager->is_gen) {
+        /* Remove existing invalid book-keeping data. */
+        hashtable_remove(&manager_table, pc);
```

Comment:
Does a lock need to be held here across this lookup and query?  Thinking of thread-private caches with two threads both creating the same block.

Reply:
There is a write lock further up. The only general issue I see with thread-private caches is that a subsequent thread is going to invalidate the book-keeping data of a first thread, thus yielding the first creation useless. Not really sure how to get around this issue though or if I should in this PR. Could be an XY problem

Reply:
In particular, this is due to the hash-table being shared amongst threads.

Reply:
Thinking about this further - basic block data only because invalidated when dynamically generating a new case (denoted by the is_gen flag) or when the code of the basic block changes due to app performing some JIT or when a library is unloaded and another is loaded. drwrap compares the  call-bytes to check consistency. I could look into more specific consistency checks to avoid unnecessary invalidation instead of relying on DR's new bb events. hmm

Reply:
Seems fine to get something working now and separate out theoretical corner cases for later.  The main short-term goal is to try to stabilize the interface for the DR 8.0 release (the only other thing blocking 8.0 is that I wanted to finish #3966 / #4200 b/c it's a compat break; I started on it but the clean call alignment is complicated and just needs more time -- could prob finish this week).

Reply:
Thanks for taking a look. I think this PR is done and ready for merge. Iill move on to the analysis API.